### PR TITLE
Add warning for checkbox with value prop but no checked prop

### PR DESCRIPTION
**Pull request by athomann**
Referencing #9446

### CODE REVIEW

**Comment by gunn** [comment]
You might want to add a test for this in ReactDOMInput-test.js :

```javascript
it('should warn for type checkbox with a value prop and no checked prop', () => {
  ReactTestUtils.renderIntoDocument(
    <input type="checkbox" checked={true} onChange={() => {}} />,
  );
  expectDev(console.error.calls.count()).toBe(0);

  ReactTestUtils.renderIntoDocument(
    <input type="checkbox" value={true} onChange={() => {}} />,
  );
  expectDev(console.error.calls.count()).toBe(1);
  expectDev(normalizeCodeLocInfo(console.error.calls.argsFor(0)[0])).toBe(
    'A component contains an input of type checkbox with a value prop, but not a checked prop. ' +
      'If you use value instead of checked with a controlled checkbox ' +
      "it looks like its working but it's not. The input remains " +
      "uncontrolled and if its value is updated in state elsewhere it won't " +
      'update in the DOM. More info: ' +
      'https://fb.me/react-controlled-components',
  );
});
```

**Comment by athomann** [comment]
Thanks, added.

File: src/renderers/dom/fiber/wrappers/ReactDOMFiberInput.js [hunk 1]
```diff
@@ -97,6 +98,39 @@ var ReactDOMInput = {
         getCurrentFiberStackAddendum,
       );
 
+      if (props.type === 'checkbox') {
+        warning(
+          false,
+          '%s contains an input of type %s with a value prop, but not a checked prop. ' +
+            'If you use value instead of checked with a controlled checkbox ' +
+            "it looks like its working but it's not. The input remains " +
```
Comment (jkascend) [inline src/renderers/dom/fiber/wrappers/ReactDOMFiberInput.js]: `it looks like its working but it's not. The input remains `

Change to "it looks like **_it's_** working but it's not."

File: src/renderers/dom/shared/wrappers/__tests__/ReactDOMInput-test.js [hunk 2]
```diff
@@ -842,6 +842,26 @@ describe('ReactDOMInput', () => {
...
+    expectDev(console.error.calls.count()).toBe(0);
+
+    ReactTestUtils.renderIntoDocument(
+      <input type="checkbox" value={true} onChange={() => {}} />,
+    );
+    expectDev(console.error.calls.count()).toBe(1);
+    expectDev(normalizeCodeLocInfo(console.error.calls.argsFor(0)[0])).toBe(
+      'Warning: A component contains an input of type checkbox with a value prop, but not a checked prop. ' +
+        'If you use value instead of checked with a controlled checkbox ' +
+        "it looks like its working but it's not. The input remains " +
```
Comment (jkascend) [inline src/renderers/dom/shared/wrappers/__tests__/ReactDOMInput-test.js]: Expect updated text from above, **_it's_**

File: src/renderers/dom/fiber/wrappers/ReactDOMFiberInput.js [hunk 3]
```diff
@@ -97,6 +98,39 @@ var ReactDOMInput = {
...
+        props.type === 'checkbox' &&
+        props.checked === undefined &&
+        props.value !== undefined &&
+        !didWarnValueChecked
+      ) {
+        warning(
+          false,
+          '%s contains an input of type %s with a value prop, but not a checked prop. ' +
+            'If you use value instead of checked with a controlled checkbox ' +
+            "it looks like its working but it's not. The input remains " +
```
Comment (jkascend) [inline src/renderers/dom/fiber/wrappers/ReactDOMFiberInput.js]: Like above, "looks like **_it's_** working"

File: src/renderers/dom/stack/client/wrappers/ReactDOMInput.js [hunk 4]
```diff
@@ -88,6 +89,26 @@ var ReactDOMInput = {
...
+        props.type === 'checkbox' &&
+        props.checked === undefined &&
+        props.value !== undefined &&
+        !didWarnValueChecked
+      ) {
+        warning(
+          false,
+          '%s contains an input of type %s with a value prop, but not a checked prop. ' +
+            'If you use value instead of checked with a controlled checkbox ' +
+            "it looks like its working but it's not. The input remains " +
```
Comment (jkascend) [inline src/renderers/dom/stack/client/wrappers/ReactDOMInput.js]: Update to "looks like **_it's_** working"

**Comment by Hypnosphi** [comment]
@jkascend you were going to merge it, but what do you think about this? https://github.com/facebook/react/issues/9446#issuecomment-296489631

It should be OK to have `value` prop in an uncontrolled checkbox, it just has another meaning.

**Comment by jkascend** [comment]
@Hypnosphi I was operating under the assumption that the reasoning behind the PR is valid. Just wanted to prevent the merging of typographical errors.

File: src/renderers/dom/fiber/wrappers/ReactDOMFiberInput.js [hunk 5]
```diff
@@ -97,6 +98,39 @@ var ReactDOMInput = {
         getCurrentFiberStackAddendum,
       );
 
+      if (props.type === 'checkbox') {
```
Comment (philipp-spiess) [inline src/renderers/dom/fiber/wrappers/ReactDOMFiberInput.js]: I don't quite understand this warning. I think the one below should be enough. Maybe this is a merge conflict of some sort?

File: src/renderers/dom/shared/wrappers/__tests__/ReactDOMInput-test.js [hunk 6]
```diff
@@ -842,6 +842,26 @@ describe('ReactDOMInput', () => {
     );
   });
 
+  it('should warn for type checkbox with a value prop and no checked prop', () => {
+    ReactTestUtils.renderIntoDocument(
+      <input type="checkbox" checked={true} onChange={() => {}} />,
+    );
+    expectDev(console.error.calls.count()).toBe(0);
```
Comment (philipp-spiess) [inline src/renderers/dom/shared/wrappers/__tests__/ReactDOMInput-test.js]: We have a helper for this now. For an example, check out: https://github.com/facebook/react/blob/725e499cfb4da60da46aa5b44c4bad29cad3d08f/packages/react-dom/src/__tests__/ReactDOMInput-test.js#L74-L80

File: src/renderers/dom/fiber/wrappers/ReactDOMFiberInput.js [hunk 7]
```diff
@@ -97,6 +98,39 @@ var ReactDOMInput = {
...
+            "uncontrolled and if its value is updated in state elsewhere it won't " +
+            'update in the DOM. More info: ' +
+            'https://fb.me/react-controlled-components',
+          getCurrentFiberOwnerName() || 'A component',
+          props.type,
+        );
+      }
+
+      if (
+        props.type === 'checkbox' &&
```
Comment (philipp-spiess) [inline src/renderers/dom/fiber/wrappers/ReactDOMFiberInput.js]: I think we want the same behavior for `radio` buttons as well.

**Pull request closed without merging**